### PR TITLE
stdenv: make meta.position overridable

### LINKED ISSUE #295907: `meta.position` specification is not overridable via `<pkg>.overrideAttrs` overriding
### Describe the bug

According to [Nix Manual](https://nixos.org/manual/nix/stable/command-ref/new-cli/nix3-edit.html?highlight=meta.position#description) and the [`stdenv.mkDerivation` source](https://github.com/NixOS/nixpkgs/blob/master/pkgs/stdenv/generic/make-derivation.nix#L516-L521), `meta.position` is specified by `stdenv.mkDerivation` according to several attributes of the package.

As `meta.position` is also used to update packages automatically, it is sometimes necessary to specify/override with a canonical value. A real-world example can be found in #295828. However, such overriding via `<pkg>.overrideAttrs` provided by `stdenv.mkDerivation` seems simply discarded.

### Steps To Reproduce
```
$ nix repl
Welcome to Nix 2.18.1. Type :? for help.

nix-repl> pkgs = import ./. { }

nix-repl> pkgs.hello.meta.position
"/home/shamrock/Projects/NixOS/nixpkgs_timeshift/pkgs/by-name/he/hello/package.nix:34"

nix-repl> pkgs.cowsay.meta.position
"/home/shamrock/Projects/NixOS/nixpkgs_timeshift/pkgs/tools/misc/cowsay/default.nix:46"

nix-repl> (pkgs.hello.overrideAttrs (previousAttrs: { meta = previousAttrs.meta // { position = pkgs.cowsay.meta.position; }; })).meta.position
"/home/shamrock/Projects/NixOS/nixpkgs_timeshift/pkgs/by-name/he/hello/package.nix:34"
```

Or
`nix eval --impure --expr "let pkgs = import ./. { }; in (pkgs.hello.overrideAttrs (previousAttrs: { meta = previousAttrs.meta // { position = pkgs.cowsay.meta.position; }; })).meta.position"`

### Expected behavior
The `meta.position` attribute be overridable with `<pkg>.overrideAttrs`, i.e.

```
$ nix eval --impure --expr "let pkgs = import ./. { }; in (pkgs.hello.overrideAttrs (previousAttrs: { meta = previousAttrs.meta // { position = pkgs.cowsay.meta.position; }; })).meta.position == pkgs.cowsay.meta.position"
true
```

### Metadata
Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```console
[user@system:~]$ nix-shell -p nix-info --run "nix-info -m"
output here
```

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

### PR DESCRIPTION
**Pull request by l0b0**
Rebase of orphaned #295973, since the original can't be merged as-is. Original description:

> this is done simply by changing the order in which attrsets are merged with //
>
> fixes #295907

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by l0b0** [comment]
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 359191`

---
### `x86_64-linux`

**Comment by emilazy** [comment]
Has the `inherit meta;` footgun discussed in https://github.com/NixOS/nixpkgs/pull/295973#issuecomment-2286598811 been addressed?

**Comment by pbsds** [comment]
Perhaps we need to support `meta.position = lib.mkForce "some value"`?

**Comment by l0b0** [comment]
@emilazy @pbsds I rebased the original branch with the assumption that it was ready to merge except for the conflict, since it already had two approvals. My only motivation was to do what looked like a small amount of work to finish something others had spent effort on and seemed to care about. So I'm afraid I don't have enough context for your comments. Would either of you be willing to take over, or to ELI5 what's needed to push this over the line?

**Comment by pbsds** [comment]
The problem is that `meta.position` will become _too easy_ to override by mistake. For example the common pattern `meta = foobar-unwrapped.meta // { homepage = xyz; }` or `inherit (foobar-unwrapped) meta` will inherit the `position` of `foobar-unwrapped`.

Only allowing to override the `meta.position` when explicitly intended would be a way to solve this. Whether we want to opt into this by wrapping the position value (e.g. `meta.position = lib.mkForce ...`) or with a flag (e.g. `meta.__enableOverridePosition = true`) can be discussed. I lean towards the latter.

**Comment by l0b0** [comment]
I mistakenly thought the only remaining issue on #295973 was rebasing onto master. Closing.

**Comment by eclairevoyant** [comment]
> The problem is that `meta.position` will become _too easy_ to override by mistake. For example the common pattern `meta = foobar-unwrapped.meta // { homepage = xyz; }` or `inherit (foobar-unwrapped) meta` will inherit the `position` of `foobar-unwrapped`.
>

@pbsds That's already the case today and this PR doesn't change that.
I don't think this PR should be responsible for correcting that, it's scope creep.

Historically I have generally corrected those who write `inherit (foobar) meta;` to explicitly inherit the attrs they care about.

**Comment by emilazy** [comment]
How is it already the case when `meta.position` is ignored on the input side?

**Comment by eclairevoyant** [comment]
I wasn't talking about overrides specifically, rather inheriting meta in general in package expressions. Due to the inherent breakage of `meta.position` it's a bad practice to inherit meta wholesale and should be corrected irrespective of this PR. Here's an example I fixed: https://github.com/NixOS/nixpkgs/pull/334036

**Comment by pbsds** [comment]
That example is verbose and a maintenance burden. I think it makes more sense with a magic `__dontClobberPosition` to let the `meta.position` go through, or even allow using `lib.mkForce`

**Comment by eclairevoyant** [comment]
Okay, great! That should still be a separate PR instead of imposing burden on contributors trying to improve another area. Perhaps it should come first, even, but still, a separate PR.

**Comment by pbsds** [comment]
I don't take the decision to block lightly. I opted to block, since merging this as-is would make https://search.nixos.org _far less_ useful.
https://gist.github.com/pbsds/ccd24294cd6b77ea0c9f48d01b39abc5 via https://github.com/NixOS/nixpkgs/pull/329900#issuecomment-2250297217 is filled with cases where the builder would set `meta.position` to its location instead of the derivation. This might however now be possible to fix for all `lib.extendMkDerivation`-based builders.

**Comment by emilazy** [comment]
> I wasn't talking about overrides specifically, rather inheriting meta in general in package expressions. Due to the inherent breakage of `meta.position` it's a bad practice to inherit meta wholesale and should be corrected irrespective of this PR. Here's an example I fixed: #334036

Okay, you’re right: I didn’t realize that already happened with `builtins.unsafeGetAttrPos "description" attrs.meta`. I sort of feel like that should just be removed, though, or at least moved below the other non‐`meta` cases. I think `inherit (pkg) meta;` is a reasonable thing to want to do and applicable in many circumstances.

@pbsds Do you know why those examples aren’t already broken? Do they not have `meta.description` set?

**Comment by pbsds** [comment]
The presence of `meta.description` seems indeed to be what determines whether `meta.position` currently is overridden or not, TIL.

This is `elf-header-real` on `master`:

```
$ nix eval --log-format raw --file default.nix elf-header-real.meta --json | jq '{position, maintainersPosition}'
{
  "position": "/home/pbsds/repos/nixpkgs/master/pkgs/by-name/el/elf-header-real/package.nix:39",
  "maintainersPosition": {
    "column": 11,
    "file": "/home/pbsds/repos/nixpkgs/master/pkgs/development/libraries/glibc/common.nix",
    "line": 345
  }
}
```

If I comment out the `meta.description`:

```
$ nix eval --log-format raw --file default.nix elf-header-real.meta --json | jq '{position, maintainersPosition}'
{
  "position": "/home/pbsds/repos/nixpkgs/master/pkgs/development/libraries/glibc/default.nix:218",
  "maintainersPosition": {
    "column": 11,
    "file": "/home/pbsds/repos/nixpkgs/master/pkgs/development/libraries/glibc/common.nix",
    "line": 345
  }
}
```

Interestingly the new `maintainersPosition` is not overridden like `position` is, meaning we now have a difference in behavior on our hands.

I don't think we should ban inheriting all attributes from `meta` without aid from CI. Otherwise it would just cause lot of toil, review cycles, and discord. 
If we want to fix this properly we could have mkDerivation add a `__dontInheritMe` to `meta` which is disallowed when `config.checkMeta=True`. To avoid the maintenance burden and verbosity of `inherit (parent.meta) ...;` we could also introduce a `inheritSafeMeta` derivation attribute. This will also make it way easier to fix all cases with a treewide before we enable the `checkMeta` check.

What do you think @emilazy @eclairevoyant?

**Comment by eclairevoyant** [comment]
> I don't think we should ban inheriting all attributes from `meta` without aid from CI. Otherwise it would just cause lot of toil, review cycles, and discord.

Fair point

> we could have mkDerivation add a `__dontInheritMe` to `meta` which is disallowed when `config.checkMeta=True`. To avoid the maintenance burden and verbosity of `inherit (parent.meta) ...;` we could also introduce a `inheritSafeMeta` derivation attribute.

Sounds like a reasonable approach
